src/FileDownloader/ChunkFileDownloader.cs

```diff
@@ -115,46 +149,37 @@ public async Task DownloadAsync(CancellationToken ct = default)
             }
             finally
             {
-                OnProgressChanged();
+                ReportProgress(null); //Forced
             }
         }
 
-        private async Task OrchestrateDownloadWorkersAsync(IEnumerable<FileRange> chunks, CancellationToken ct)
+        private async Task DequeueAndDownloadChunksAsync(CancellationToken ct = default)
         {
-            var downloaderWorkers = new ConcurrentQueue<Task>();
-            _chunkDownloadQueue = new ConcurrentQueue<FileRange>(chunks);
-
-            while (_activeDownloads < MaxConcurrentDownloads && _chunkDownloadQueue.Count > 0 && !_downloadFailed)
+            while (_chunkDownloadQueue.TryDequeue(out var fileChunk))
             {
-                _activeDownloads++;
+                var workerFailedDownloadingOnce = false;
 
-                downloaderWorkers.Enqueue(Task.Run(DequeueAndDownloadChunksAsync, ct));
-                await Task.Delay(1000, ct);
-            }
-
-            await Task.WhenAll(downloaderWorkers);
-        }
-
-        private async Task DequeueAndDownloadChunksAsync()
-        {
-            var workerFailedDownloading = false;
-            var taskId = _activeDownloads;
-
-            while (_chunkDownloadQueue.TryDequeue(out var fileChunk) && !workerFailedDownloading)
-            {
+                retry:
                 var chunkDownload = new ChunkDownload(DownloadUrl, fileChunk, _downloadTempFolder);
-                var downloadSuccesfullyCompleted = await chunkDownload.TryDownloadAsync(IncrementTotalDownloadProgress);
+                var downloadSuccessfullyCompleted =
+                    await chunkDownload.TryDownloadAsync(IncrementTotalDownloadProgress, ct);
 
-                if (!downloadSuccesfullyCompleted)
-                {
-                    _chunkDownloadQueue.Enqueue(fileChunk);
-                    workerFailedDownloading = true;
-                    _downloadFailed = true;
-                }
-                else
+                if (!downloadSuccessfullyCompleted)
                 {
-                    _completedChunks.TryAdd(fileChunk.Start, chunkDownload.DownloadTmpFileName);
+                    if (workerFailedDownloadingOnce)
+                    {
+                        _chunkDownloadQueue.Enqueue(fileChunk);
+                        break;
+                    }
+
+                    workerFailedDownloadingOnce = true;
+                    await Task.Delay(1000, ct);
+                    goto retry;
```

Comment:
👎 do not use go-tos

Reply:
I am not going to explain why this is a terrible idea to do, there is plenty of information about this on the net.